runsc/container/container.go

```diff
@@ -951,41 +989,48 @@ func (c *Container) createGoferFilestores(ovlConf config.Overlay2, mountHints *b
 	return goferFilestores, goferConfs, nil
 }
 
-func (c *Container) createGoferFilestore(ovlConf config.Overlay2, mountSrc string, shouldOverlay bool, hint *boot.MountHint) (*os.File, boot.GoferMountConf, error) {
-	// MountHint information takes precedence over shouldOverlay.
-	if hint != nil && !specutils.IsGoferMount(hint.Mount) {
-		switch hint.Mount.Type {
-		case "tmpfs":
-			// Create self-backed tmpfs.
-			return c.createGoferFilestoreInSelf(mountSrc, hint, boot.SelfTmpfs)
-		default:
-			return nil, boot.VanillaGofer, fmt.Errorf("unsupported mount type %q in mount hint", hint.Mount.Type)
-		}
-	}
-	switch {
-	case !shouldOverlay:
-		return nil, boot.VanillaGofer, nil
-	case ovlConf.IsBackedByMemory():
-		return nil, boot.MemoryOverlay, nil
-	case ovlConf.IsBackedBySelf():
-		return c.createGoferFilestoreInSelf(mountSrc, hint, boot.SelfOverlay)
+func (c *Container) createGoferFilestore(overlayMedium config.OverlayMedium, mountSrc string, mountType string, isShared bool) (*os.File, boot.GoferMountConf, error) {
+	var lower boot.GoferMountConfLowerType
+	switch mountType {
+	case boot.Bind:
+		lower = boot.Lisafs
+	case tmpfs.Name:
+		if overlayMedium == config.NoOverlay {
+			overlayMedium = config.SelfOverlay
+		}
```

Comment:
Ah I see, good catch. So this is only possible when some container is mounting the EmptyDir as a read-only mount in one container. We still want to create the self filestore (in case some other container mounts it writably, the shared mount should be created correctly). This detail is important. Instead of handling this as a special case here, we should just move lines 972-974 above line 965:
```
		if specutils.IsReadonlyMount(c.Spec.Mounts[i].Options) {
			overlayMedium = config.NoOverlay
		}
		if hint := mountHints.FindMount(c.Spec.Mounts[i].Source); hint != nil {
			// Note that we want overlayMedium=self even if this is a read-only mount so that
			// the shared mount is created correctly.
			overlayMedium = config.SelfOverlay
			if !specutils.IsGoferMount(hint.Mount) {
				mountType = hint.Mount.Type
			}
			isShared = hint.ShouldShareMount()
		}
```

Reply:
Makes sense. Will do. Thanks!